### ISSUE TITLE
RTP playback issues

### ISSUE THREAD

**Issue by kanavin** [01org/wysiwidi#60]
Try playing streams with something like 
```
GST_DEBUG_DUMP_DOT_DIR=. gst-launch-1.0 udpsrc port=5000 caps="application/x-rtp" ! rtpmp2tdepay ! decodebin name=decoder ! "audio/x-raw" ! .audio_sink playsink name=sink    decoder. ! "video/x-raw" ! sink.video_sink
```
(and also with jitterbuffer in there)

**Comment by jku** [comment]
Some results from the testing: The streaming seems to work fine if there are no UDP packet problems (missing packets or reordering). When it fails I can typically see lot of messages from jitterbuffer about dropping packets that arrived too late and then a couple of decoder errors. After this the playback shows a frame here and a frame there and never really recovers. Possibly the av synchronization fails after that as using "sync=false" on the sinks makes playback keep somewhat working after problems (but then it's badly out of sync).
 
I was talking to Sreerenj again after he testing. The take away I got was that normally in gstreamer the elements before the decoder must guarantee that the decoder gets all frames in the correct order: there's no requirement for the decoder to cope with missing frames. So I'm guessing this means two things:
 * for the (hopefully rare) case where we just don't have the packets, we should "restart" the pipeline -- maybe seek to the next keyframe we get or something?
 * for the other cases, we really need the retransmission requests...

So I spent  while trying to understand rtpbin and/or rtpsession to get the rtcp handling. I was hoping something like this would work (this is still without the rtcp parts to keep it simple):

    [ wait, adding this in a moment... ]

but it just says "streaming task paused, reason not-negotiated (-4). Any ideas?


Also Sreerenj said he could spend some time next week looking at this if haven't figured it out.

**Comment by jku** [comment]
Also, I think this is a way to sort of reproduce this problem without all the P2P/Miracast: just before the udpsink element add a "identity drop-probability=0.05"  element

**Comment by kanavin** [comment]
Yay for easy reproduction!

I'd say we should also talk to the authors of jitterbuffer. Lost packets that we simply will never get are a normal thing in RTP, so there has to be a mechanism for dealing with that, that doesn't ask too much from the app developers (like, 'restarting the pipeline' or anything of the sort).

I have never been able to get a working pipeline out of rtpbin (didn't bother with rtpsession). If you want to see exactly what's wrong, get a graphical representation of the pipeline - you'll probably see that some element wants to connect an RTCP pad, and doesn't have anything to connect it to.

**Comment by jku** [comment]
The mechanism to deal with packet loss in RTP is retransmission requests in RTCP, right? So maybe we just need to get that working and see if that hopefully deals with 99% of the problems...

The rtpsession element has pads  for RTCP but it should work without them: I was hoping to get the minimal example working first... In any case, here's how I think it should work, for the same audio only example:

    gst-launch-1.0 audiotestsrc wave=0 ! avenc_ac3 ! mpegtsmux ! rtpmp2tpay ! .send_rtp_sink rtpsession name=session .send_rtp_src ! udpsink host=127.0.0.1 port=5000 session.send_rtcp_src ! udpsink host=127.0.0.1 port=5001
    
    gst-launch-1.0 udpsrc port=5000 caps="application/x-rtp,clock-rate=(int)90000" ! .recv_rtp_sink rtpsession name=session .recv_rtp_src ! rtpmp2tdepay ! decodebin ! autoaudiosink udpsrc port=5001 caps="application/x-rtcp" ! session.recv_rtcp_sink

**Comment by kanavin** [comment]
As far as WFD devices go, RTCP is optional on both sides. The sink can (but doesn't have to) propose to the source to use it, and the source can (but doesn't have to) accept the proposal. So we still need to figure out the RTCP-less scenario.

Otherwise, yes, it would be interesting to see if things improve significantly with RTCP in use.

**Comment by jku** [comment]
Well, agreed. But two points:
* Android claims to support RTCP (I checked). This is pretty important in terms of demoability as well as actual interoperability: I'm guessing most Mircast sources are android by far.
* If the decoders cannot handle missing packets, then we have to do _something_ based on the rtp element signals before the decoder tries to decode the broken part of the stream -- this is how I understood the situation: maybe "something" means seeking to the next seekable frame?

**Comment by jku** [comment]
My RtpBin test setup so far (doesn't work yet, just documenting here):

    # sender
    gst-launch-1.0 -v rtpbin name=rtp_bin \
                      udpsrc name=rtcp_udp_src port=5002 \
                          ! rtp_bin.recv_rtcp_sink_0 \
                      audiotestsrc wave=0 \
                          ! avenc_ac3 \
                              ! mpegtsmux \
                                  ! rtpmp2tpay \
                                      ! rtp_bin.send_rtp_sink_0 \
                      rtp_bin.send_rtcp_src_0 \
                          ! udpsink name=rtcp_udp_sink host=127.0.0.1 port=5001 sync=false async=false \
                      rtp_bin.send_rtp_src_0 \
                          ! udpsink name=rtp_udp_sink host=127.0.0.1 port=5000

    # receiver
    gst-launch-1.0 -v rtpbin name=rtp_bin \
                      udpsrc name=rtp_udp_src port=5000 caps="application/x-rtp,clock-rate=(int)90000,payload=(int)33,encoding-name=(string)MP2T" \
                          ! rtp_bin.recv_rtp_sink_0 \
                      udpsrc name=rtcp_udp_src port=5001 \
                          ! rtp_bin.recv_rtcp_sink_0 \
                      rtp_bin. \
                          ! rtpmp2tdepay \
                              ! decodebin \
                                  ! autoaudiosink \
                      rtp_bin.send_rtcp_src_0 \
                          ! udpsink name=rtcp_udp_sink host=127.0.0.1 port=5002 sync=false async=false

**Comment by kanavin** [comment]
I have found why playback breaks down when packets are lost and filed a bug upstream:
https://bugzilla.gnome.org/show_bug.cgi?id=743740

**Comment by kanavin** [comment]
The fix for the issue is available here:
https://bugzilla.gnome.org/show_bug.cgi?id=743363#c7

**Comment by kanavin** [comment]
Fix is now available in 1.4 and 1.5 branches of gstreamer, closing.
